### PR TITLE
Revision/expanding sequences Sequences with default value get not expanded

### PR DESCRIPTION
**Pull request by henhuy**
### This revision aims to reduce unnecessary list extensions while creating the model - see related issue #436. 

**Related issue**: #436 
**Located in**: Major change was made to `oemof.solph.plumbing`
**Functionality**:
While building the sequence an internal attribute `self.default_changed` is set to False. Whenever `__getitem__` of sequence is called, only the default value is returned. Before, every time an internal list was extended to requested index. Only, if `__setitem__` is called attribute `self.default_changed` is set to True and an internal list is build, as the sequence no longer consists of only one default value.
**Outcome**:
This revision leads to three things:
1. Testing run-time decreased! (for me, from 8sec to 5sec!) :)
2. `__len__` function of Sequences with only one default value now returns 0 - this is no problem for all tests, but users looking into solph components may get different results.
3. I changed param_results-API, so that Sequences containing only the default value are now added to the 'scalars' dictionary.

**Not implemented**:
The other idea from #436, setting up a list containing entries for every time-step is not implemented and maybe even unnecessary, due to new implementation!

I think these changes are good, improving oemof solph! What do you think?

### CODE REVIEW

**Comment by henhuy** [comment]
As the param_results dict was changed (to the good, I think) this will have implications for #433. If this PR is accepted, I will change related calculations...

**Comment by henhuy** [comment]
I wonder why there are no travis checks for this PR? Can you explain me?

File: oemof/solph/plumbing.py [hunk 1]
```diff
@@ -65,16 +65,20 @@ class _Sequence(UserList):
     """
     def __init__(self, *args, **kwargs):
         self.default = kwargs["default"]
+        self.default_changed = False
         super().__init__(*args)
 
     def __getitem__(self, key):
+        if not self.default_changed:
+            return self.default
```
Comment (simnh) [inline oemof/solph/plumbing.py]: Do I understand this right, that no iterable object but a scalar is returned, if ` not default_changed`? I think, it kind of strange to have a sequence class that does not return a iterable object. My idea was rather to move this to the flow/node classed instead.

**Comment by simnh** [comment]
Sorry, I think I did not read correctly...Nevertheless I am not sure, whether we should mix the idea of default values with the sequence class..

**Comment by uvchik** [comment]
> I wonder why there are no travis checks for this PR? Can you explain me?

That's strange because the test has been done: https://travis-ci.org/oemof/oemof/builds/329396739

The test failed by the way :-P

**Comment by henhuy** [comment]
Bit confused @simnh : Is your request still open, or did you change your mind and approved?

> Nevertheless I am not sure, whether we should mix the idea of default values with the sequence class..

I don't really see the advantage of creating a variable length list, if `solph` only needs the default value every time...
> The test failed by the way :-P

Sorry, i did not check for the docstring tests (I will change my testing parameters to take these into account for the future). The test fails, because no iterable list is created internally. 

So, there is an decision to make, if this behaviour is wanted or not. Personally, I prefer my implementation as oemof is set up faster and I don't see the point in creating a list of 8760 times the same value... 
But no offense there - it was just an idea!

**Comment by henhuy** [comment]
I fixed sequences __len__, __setitem__ and __iter__ function in order to emulate typical list behavior without actually creating list internally. 
Maybe this resolves your last concerns @simnh?

**Comment by henhuy** [comment]
What do you mean with `plumbing` is not in `core` atm?

**Comment by simnh** [comment]
> Is your request still open, or did you change your mind and approved?

I already approved this one, as I changed my mind ;-)

> What do you mean with plumbing is not in core atm?

the module is not located on the core level but in the solph package...because you said above 'this will improve oemof core'

**Comment by henhuy** [comment]
> plumbing is not in core atm...
Ah okay! (I had to re-read my own text) Thanks for the hint - changed PR description...

**Pull request merged**